nvaccess/nvda
# MS Word with UIA: heading quicknav no longer keeps repeating the last heading

### Link to issue number:
Fixes #9540

### Summary of the issue:
When accessing MS Word documents with UIA, if the last bit of text is formatted as a heading, pressing quicknav by heading (h) in browse mode keeps repeating that heding. Similarly, The elements list when set to show headings, lists the final heading twice.

This is due to the fact that most UIA providers when on the last unit, move to the collapsed exclusive end if asked to move forward 1 by that unit. Only after that does the move fail.
We could add some specific checks in the UIABrowseMode.UIAHeadingQuicknavIterator function to take this into account, or we can rewrite that algorithm to make use of iterUIARangeByUnit which is used in many other places and avoids that bug all together.
 
### Description of how this pull request fixes the issue:
Rewrite the UIABrowseMode.UIAHeadingQuicknavIterator function to make use of iterUIARangeByUnit. It is now much more similar to the UIABrowseMode.UIATextAttributeQuicknavIterator function, which we know does not have the repeating bug.
  
### Testing strategy:
* Ensure NVDA's use UIA to access MS Word document controls when available is on
* Open a new blank document in MS Word
* Type `Test` and press enter
* Press alt+shift+rightArrow to start formatting as heading 2 style and type `First heading` and press enter
* Type `Test` and press enter
* Press alt+shift+rightArrow to start formatting as heading 3, type `Second heading` and press enter 
* `Second heading` should be the final line of text in the document, I.e. the heading ends at the end of the document.
* Switch to browse mode with NVDA+space
* Move to the top of the document with control+home
* Press h to jump to the `First heading`
* Press h to jump to the `Second heading`
* Press h and ensure that NVDA reports that there is no next heading.
* Press shift+h to move to the `First heading`
* Press shift+h and ensure NVDA reports that there is no previous heading
* Press NVDA+f7 to open the NVDA elements list, and select Headings from the Type radio buttons
* Ensure that just two headings are listed in the treeview: `First heading` and `Second heading`.

### Known issues with pull request:
None known.

### Change log entries:
New features
Changes
Bug fixes
* MS word with UIA: heading quick nav in browse mode no longer gets stuck on the final heading of a document, nor is this heading shown twice in the NVDA elements list.

For Developers

### Code Review Checklist:



  - description is up to date
  - change log entries
  - Unit tests
  - System (end to end) tests
  - Manual testing
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
